profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/internal/BaseProfitBricksLiveTest.java

```diff
@@ -16,16 +16,135 @@
+      );
+   
+      ComputeConstants c = computeConstants;
+      
+      Predicate<String> serverAvailableCheck = new Predicate<String>() {
+         @Override
+         public boolean apply(String complexId) {
+            String[] ids = complexId.split(",");
+
+            Server server = api.serverApi().getServer(ids[0], ids[1]);
+
+            if (server == null || server.metadata().state() == null)
+               return false;
+
+            return server.metadata().state() == State.AVAILABLE;
+         }
+      };
+      
+      serverAvailable = Predicates2.retry(serverAvailableCheck, c.pollTimeout(), c.pollPeriod(), c.pollMaxPeriod(), TimeUnit.SECONDS);
+      
+      Predicate<String> serverRemovedPredicate = new Predicate<String>() {
+         @Override
+         public boolean apply(String complexId) {
+            String[] ids = complexId.split(",");
+            return api.serverApi().getServer(ids[0], ids[1]) == null;
+         }
+      };
+      
+      serverRemoved = Predicates2.retry(serverRemovedPredicate, c.pollTimeout(), c.pollPeriod(), c.pollMaxPeriod(), TimeUnit.SECONDS);
+
+      return injector.getInstance(ProfitBricksApi.class);
+   }
+   
+   protected void assertRandom(Predicate<String> check, String arguments) {
+      ComputeConstants c = computeConstants;
+      Predicate<String> checkPoll = Predicates2.retry(check, c.pollTimeout(), c.pollPeriod(), c.pollMaxPeriod(), TimeUnit.SECONDS);
+      assertTrue(checkPoll.apply(arguments), "Random check failed in the configured timeout");
+   }
+
+   protected void assertDataCenterAvailable(DataCenter dataCenter) {
+      assertDataCenterAvailable(dataCenter.id());
+   }
+
+   protected void assertDataCenterAvailable(String dataCenterId) {
+      assertTrue(dataCenterAvailable.apply(dataCenterId),
+              String.format("Datacenter %s wasn't available in the configured timeout", dataCenterId));
+   }
+
+   protected void assertNodeRunning(String serverId) {
+      assertTrue(serverRunning.apply(serverId), String.format("Server %s did not start in the configured timeout", serverId));
+   }
+
+   protected void assertNodeSuspended(String serverId) {
+      assertTrue(serverSuspended.apply(serverId), String.format("Server %s did not stop in the configured timeout", serverId));
+   }
+   
+   protected void assertNodeRemoved(String serverId) {
+      assertTrue(serverRemoved.apply(serverId), String.format("Server %s was not removed in the configured timeout", serverId));
+   }
+   
+   protected void assertNodeAvailable(String serverId) {
+      assertTrue(serverAvailable.apply(serverId), String.format("Server %s is not available", serverId));
+   }
+   
+   protected DataCenter createDataCenter() {
+      return api.dataCenterApi().create("test-data-center", "example description", TestLocation.value());
+   }
+   
+   protected void deleteDataCenter(String id) {
+      api.dataCenterApi().delete(id);
+   }
+   
+   protected String complexId(String ... ids) {
```

Comment:
This `complexId` thing is horrible. It is coupling the codebase with the fact that you concatenate IDs for convenience. For example, the predicates defined in the predicates module have to split the input string, just because "they know" it is the concatenation of two IDs.

Instead of coupling concrete behavior in the entire codebase, and doing concatenations and splits here and there, create a proper `ServerRef` or `DatacenterAndServer`, or whatever, that you can pass between application layers, and provides convenient constructors and accessors for the internal IDs. This way the predicate remains clean, and the entire codebase remains clean and the "id encoding" logic properly encapsulated.